pkg/project/auth/cache.go

```diff
@@ -0,0 +1,482 @@
+	addSubjectsToNamespace(groupSubjectRecordStore, review.Groups(), namespace)
+	cacheReviewRecord(request, lastKnownValue, review, reviewRecordStore)
+	return nil
+}
+
+// List returns the set of namespace names the user has access to view
+func (ac *AuthorizationCache) List(userInfo user.Info) (*kapi.NamespaceList, error) {
+	keys := util.StringSet{}
+	user := userInfo.GetName()
+	groups := userInfo.GetGroups()
+
+	obj, exists, _ := ac.userSubjectRecordStore.GetByKey(user)
+	if exists {
+		subjectRecord := obj.(*subjectRecord)
+		keys.Insert(subjectRecord.namespaces.List()...)
+	}
+
+	for _, group := range groups {
+		obj, exists, _ := ac.groupSubjectRecordStore.GetByKey(group)
+		if exists {
+			subjectRecord := obj.(*subjectRecord)
+			keys.Insert(subjectRecord.namespaces.List()...)
+		}
+	}
+
+	namespaceList := &kapi.NamespaceList{}
+	for key := range keys {
+		namespace, exists, err := ac.namespaceStore.GetByKey(key)
+		if err != nil {
+			return nil, err
+		}
+		if exists {
+			namespaceList.Items = append(namespaceList.Items, *namespace.(*kapi.Namespace))
+		}
+	}
+	return namespaceList, nil
+}
+
+// skipReview returns true if the request was satisfied by the lastKnown
+func skipReview(request *reviewRequest, lastKnownValue *reviewRecord) bool {
+
+	// if your request is nil, you have no reason to make a review
+	if request == nil {
+		return true
+	}
+
+	// if you know nothing from a prior review, you better make a request
+	if lastKnownValue == nil {
+		return false
+	}
+	// if you are asking about a specific namespace, and you think you knew about a different one, you better check again
+	if request.namespace != lastKnownValue.namespace {
+		return false
+	}
+
+	// if you are making your request relative to a specific resource version, only make it if its different
+	if len(request.namespaceResourceVersion) > 0 && request.namespaceResourceVersion != lastKnownValue.namespaceResourceVersion {
+		return false
+	}
+
+	// if you see a new policy binding, or a newer version, we need to do a review
+	for k, v := range request.policyBindingUIDToResourceVersion {
+		oldValue, exists := lastKnownValue.policyBindingUIDToResourceVersion[k]
+		if !exists || v != oldValue {
+			return false
+		}
+	}
+
+	// if you see a new policy, or a newer version, we need to do a review
+	for k, v := range request.policyUIDToResourceVersion {
+		oldValue, exists := lastKnownValue.policyUIDToResourceVersion[k]
+		if !exists || v != oldValue {
+			return false
+		}
+	}
+	return true
+}
+
+// deleteSubjectsToNamespace removes the namespace from each subject
+// if no other namespaces are active to that subject, it will also delete the subject from the cache entirely
+func deleteSubjectsToNamespace(subjectRecordStore cache.Store, subjects []string, namespace string) {
+	for _, subject := range subjects {
+		obj, exists, _ := subjectRecordStore.GetByKey(subject)
+		if exists {
+			subjectRecord := obj.(*subjectRecord)
+			subjectRecord.namespaces.Delete(namespace)
```

Comment:
That's new to me.  I was just using what's exposed on the object rather than what its aliased to since that could change.